### PR TITLE
DMP-4358: Add buffer time in CaseExpiryDeletion and AssociatedObjectDataExpiryDeletion

### PR DESCRIPTION
**Pull request by Ben-Edwards-cgi**
### Change description ###
As part of DMP-3636 and DMP-2543, any case and its associated object will be Deleted as part of Case Expiry cleanup process in  DARTS. 

There is a requirement to add a Buffer time of 24 hours before the automated job deletes the object from DARTS.
 # In application.yaml add 2 properties to control the buffer time:
 ** case-expiry-buffer-duration: 24h
 ** associated-object-expiry-buffer-duration: 24h
 # In CaseExpiryDeletionAutomatedTask, while fetching the cases to be expired, add an additional criteria to make sure the case retention date is older than 24 hours (case-expiry-buffer-hours)
 # In AssociatedObjectDataExpiryDeletionAutomatedTask, add the following criteria while fetching the data from external_object_directory table for individual object - Media, Transcription document, Annotation document and Case document,
 ** add an additional criteria to make sure the retain_until_ts of the object is older than 24 hours (associated-object-expiry-buffer-hours)
 ** add a check if the object's retain_until_ts matches the object's event_dt minus 100 years (event-date-adjustment-years from application.yaml) in external_object_directory table for ARM record i.e. the event_date for the object is calculated and its same as object's retain_until_ts before DELETING the object.

*Acceptance Criteria*

*AC 1:* *CaseExpiry - more than 24 hours buffer time*

GIVEN

data can only be retained for as long as it is required to be kept for

WHEN

a case passes its retention date and buffer time is more than 24 hours i.e. retention date of the case is more than 24 hours in past

THEN

all Personally Identifiable Information of citizens on structured data must be deleted from the case and replaced with UUID text

 

*AC 2:* *CaseExpiry - less than 24 hours buffer time*

GIVEN

data can only be retained for as long as it is required to be kept for

WHEN

a case passes its retention date and buffer time is less than 24 hours i.e. retention date of the case is NOT more than 24 hours in past

THEN

all Personally Identifiable Information of citizens on structured data must *NOT* be deleted from the case and replaced with UUID text

 

*AC3:* *Associated Object - more than 24 hours buffer time and Event Date exist* 

GIVEN

data can only be retained for as long as it is required to be kept for

WHEN

a case's retention period is passed and buffer time is more than 24 hours i.e. retention date of the object is more than 24 hours in past

AND 

Event Date in external_object_directory table exist 

THEN

all the unstructured data records associated with that case must be deleted from DARTS data stores

 

*AC 4:* *Associated Object* *- less than 24 hours buffer time* *and Event Date exist*

GIVEN

data can only be retained for as long as it is required to be kept for

WHEN

a object's retention period is passed and buffer time is less than 24 hours i.e. retention date of the object is more than 24 hours in past

AND 

Event Date in external_object_directory table exist 

THEN

NO unstructured data records associated with that case gets deleted from DARTS data stores

 

*AC5:* *Associated Object - more than 24 hours buffer time and Event Date doesn't exist*

GIVEN

data can only be retained for as long as it is required to be kept for

WHEN

a object's retention period is passed and buffer time is more than 24 hours i.e. retention date of the object is more than 24 hours in past

AND 

Event Date in external_object_directory table doesn't exist 

THEN

NO unstructured data records associated with that case gets deleted from DARTS data stores

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

**Review by hemantasharma1129 — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/darts/task/runner/impl/AbstractLockableAutomatedTask.java [hunk 1]
```diff
@@ -36,7 +36,7 @@
 
 
 @Slf4j
-public abstract class AbstractLockableAutomatedTask implements AutomatedTask, AutoloadingAutomatedTask {
+public abstract class AbstractLockableAutomatedTask<T extends AbstractAutomatedTaskConfig> implements AutomatedTask, AutoloadingAutomatedTask {
```
Comment (karen-hedges) [inline src/main/java/uk/gov/hmcts/darts/task/runner/impl/AbstractLockableAutomatedTask.java:39]: Aren't you overly complicating this by task instead of just having an interface that returns the config?

File: src/main/java/uk/gov/hmcts/darts/task/runner/impl/AssociatedObjectDataExpiryDeletionAutomatedTask.java [hunk 2]
```diff
@@ -63,8 +67,9 @@ public AssociatedObjectDataExpiryDeletionAutomatedTask(
         ExternalObjectDirectoryRepository externalObjectDirectoryRepository,
         ExternalInboundDataStoreDeleter inboundDeleter,
         ExternalUnstructuredDataStoreDeleter unstructuredDeleter,
-        AuditApi auditApi) {
-
+        AuditApi auditApi,
+        @Value("${darts.storage.arm.event-date-adjustment-years}")
+        Integer eventDateAdjustmentYears) {
```
Comment (jackmaloney) [inline src/main/java/uk/gov/hmcts/darts/task/runner/impl/AssociatedObjectDataExpiryDeletionAutomatedTask.java:72]: How is this related to the buffer time ticket?

File: src/main/java/uk/gov/hmcts/darts/task/runner/impl/AssociatedObjectDataExpiryDeletionAutomatedTask.java [hunk 3]
```diff
@@ -168,6 +177,32 @@ <T extends SoftDelete & HasIntegerId> void deleteExternalObjectDirectoryEntity(
...
 
+    <T extends SoftDelete & HasIntegerId & HasRetention & CanReturnExternalObjectDirectoryEntities> boolean shouldDeleteFilter(T entity) {
+        ExternalObjectDirectoryEntity armExternalObjectDirectoryEntity = entity.getExternalObjectDirectoryEntities()
+            .stream()
+            .filter(e -> ExternalLocationTypeEnum.ARM.getId().equals(e.getExternalLocationType().getId()))
+            .findFirst()
+            .orElse(null);
+
+        if (armExternalObjectDirectoryEntity == null) {
+            log.debug("Skipping deletion of {} with id {} as there is no ARM external object directory entity",
```
Comment (jackmaloney) [inline src/main/java/uk/gov/hmcts/darts/task/runner/impl/AssociatedObjectDataExpiryDeletionAutomatedTask.java:188]: Should these be info statements? If these are skipped then you probably want to know about it.

File: src/main/java/uk/gov/hmcts/darts/task/runner/impl/AbstractLockableAutomatedTask.java (hunk 1 above)
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/task/runner/impl/AbstractLockableAutomatedTask.java:39]: @hemantasharma1129  what is your view?

Comment (hemantasharma1129) [reply src/main/java/uk/gov/hmcts/darts/task/runner/impl/AbstractLockableAutomatedTask.java:39]: I have no issue on using Generics but does it not changes the scope of the ticket? Even though its not a big change of introducing Generics but does our test suite retests all the jobs?

Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/task/runner/impl/AbstractLockableAutomatedTask.java:39]: Our existing regression suite should cover the assurance from adding in generics, and the risk of this breaking anything is very low as any issues should be identified at compile time instead of runtime

**Review by jackmaloney — APPROVED** [review]
